### PR TITLE
feat: add polora optimizer

### PR DESCRIPTION
**Pull request by NanoCode012**
# Description

https://nikhilgsh.github.io/polora/

## How has this been tested?

2x A40, Llama-3.2-1B, LoRA all-linear, `nvidia/OpenMathInstruct-2`, packed seq 2048,
**32768 tokens per optimizer step** (the paper's batch), bf16, constant LR, no warmup.

**Per-step overhead** (paper claims "at most 3%"):

| `lora_r` | overhead |
|---|---|
| 32 | **+1.6%** |
| 256 | +4.0% |

**Steps to loss**, both optimizers LR-swept with interior (bracketed) optima, 1500-step
horizon:

| AdamW | final eval | | PoLoRA | final eval |
|---|---|---|---|---|
| 1e-4 | 0.4960 | | 1e-2 | 0.4621 |
| **3e-4** | **0.4686** | | **3e-2** | **0.4556** |
| 1e-3 | 0.5087 | | 1e-1 | 0.4788 |

PoLoRA reaches AdamW's best-LR final loss at step 1076 of 1500, i.e. **1.39x fewer steps**,
inside the paper's claimed 1.20-1.74x (they report 1.48x for math at r=32). Net **~1.37x
wall-clock**.

**Memory**, same runs (peak over 60 steps, single A40):

| `lora_r` | optimizer | peak reserved | peak active |
|---|---|---|---|
| 32 | adamw_torch | 26.20 GiB | 24.05 GiB |
| 32 | polora | **26.10 GiB** | **23.97 GiB** |
| 256 | adamw_torch | 28.79 GiB | 26.70 GiB |
| 256 | polora | **28.17 GiB** | **26.03 GiB** |

## AI Usage Disclaimer

Claude

## Summary by CodeRabbit

* **New Features**
  * Added support for the PoLoRA optimizer for LoRA and QLoRA training.
  * Added configuration validation, supported hyperparameters, and compatibility with single-GPU, DDP, and FSDP2 setups.
  * Added checkpointing and sharded-training support for PoLoRA.

* **Documentation**
  * Added PoLoRA usage guidance, installation requirements, configuration details, and support-matrix entries.

* **Bug Fixes**
  * Added clear validation errors for unsupported training configurations and missing gradient requirements.

* **Tests**
  * Added coverage for optimizer configuration, validation rules, and multi-GPU FSDP2 behavior.

### CODE REVIEW

File: src/axolotl/utils/schemas/validation.py [hunk 1]
```diff
@@ -903,6 +902,74 @@ def check_muon_deepspeed_fsdp(cls, data):
...
+            data.get("deepspeed")
+            or data.get("fsdp")
+            or data.get("fsdp_config")
+            or (data.get("tensor_parallel_size") or 1) > 1
+        ):
+            # The update stacks pairs by shape and rebuilds curvature from whole weights.
+            raise ValueError(
+                "polora is not compatible with DeepSpeed, FSDP, or tensor parallelism. "
+                "Use single-GPU or DDP."
+            )
```
Comment (winglian) [inline src/axolotl/utils/schemas/validation.py]: PoLoRA is less useful without any distributed training support
